Doc: updated distributors page (#13018)
grafana/mimir: docs/sources/mimir/references/architecture/components/distributor.md

Internally, these limits are implemented using a per-distributor local rate limiter.
The local rate limiter for each distributor is configured with a limit of `limit / N`, where `N` is the number of healthy distributor replicas.
The distributor automatically adjusts the request and ingestion rate limits if the number of distributor replicas change.

> The distributor rate limits are implemented using a per-distributor local rate limiter, they require that write requests are [evenly distributed across the pool of distributors](#load-balancing-across-distributors).

Reviewer: ```suggestion
This design uses a per-distributor local rate limiter and requires that write requests be [evenly distributed across the pool of distributors](#load-balancing-across-distributors).
```

> ### Rate limiting configuration

Reviewer: ```suggestion
### Configure rate limits
```

By default, Prometheus remote write doesn't retry requests on 429 HTTP response status code. To modify this behavior, use `retry_on_http_429: true` in the Prometheus [`remote_write` configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write).
{{< /admonition >}}

### Distributors' hash ring for rate limiting

Since distributor rate limits are implemented locally, each distributor must know how many healthy distributors are running in the Mimir cluster.
To achieve this, each distributor joins a hash ring used for service discovery, which tracks the number of healthy distributor instances.

> The default configuration uses `memberlist` as backend for the distributors ring.

Reviewer: ```suggestion
By default, the distributors’ ring uses `memberlist` as its backend.
```

If you want to configure a different backend, for example, `consul` or `etcd`, you can use the following CLI flags (and their respective YAML configuration options) to configure the distributors ring KV store:

- `-distributor.ring.store`: The backend storage to use.
- `-distributor.ring.consul.*`: The Consul client configuration. Set this flag only if `consul` is the configured backend storage.
- `-distributor.ring.etcd.*`: The etcd client configuration. Set this flag only if `etcd` is the configured backend storage.

For more information, refer to [Configure hash rings](https://grafana.com/docs/mimir/<MIMIR_VERSION>/configure/configure-hash-rings/).

## Load balancing across distributors

We recommend uniformly distributing write requests across all distributor instances by placing a load balancer in front of them.
The preferred approach is a Layer 7 load balancer, which balances individual HTTP requests across distributors.

{{< admonition type="note" >}}

> If you run Grafana Mimir in a Kubernetes cluster and use a Kubernetes [Service](https://kubernetes.io/docs/concepts/services-networking/service/) as the ingress for distributors, note that a Kubernetes Service balances TCP connections across endpoints but does not balance HTTP requests within a single TCP connection.

Reviewer: ```suggestion
If you run Grafana Mimir in a Kubernetes cluster and use a Kubernetes [Service](https://kubernetes.io/docs/concepts/services-networking/service/) as the ingress for distributors, a Kubernetes Service balances TCP connections across endpoints but doesn't balance HTTP requests within a single TCP connection.
```

If you enable HTTP persistent connections, also known as HTTP keep-alive, Prometheus reuses the same TCP connection for each remote-write HTTP request of a remote-write shard.
This can cause distributors to receive an uneven distribution of remote-write HTTP requests.

To improve the balancing of requests between distributors, consider increasing `min_shards` in the Prometheus [`remote_write` configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write) block.
{{< /admonition >}}

## High-availability tracker

Remote write agents, such as Prometheus or Grafana Alloy, can be configured in pairs, which means that metrics continue to be scraped and written to Grafana Mimir even when one of the remote write agents is down for maintenance or is unavailable due to a failure.
We refer to this configuration as high-availability (HA) pairs.

The distributor includes an HA tracker.
When the HA tracker is enabled, the distributor deduplicates incoming series from Prometheus HA pairs.

> This enables you to have multiple HA replicas of the same Prometheus servers that write the same series to Mimir and then deduplicates the series in the Mimir distributor.

Reviewer: ```suggestion
This enables you to have multiple Prometheus HA replicas of the same server writing identical series to Mimir, which the distributor then deduplicates.
```
